Update Adding interactivity tutorial to new style (#39157)
mdn/content: files/en-us/learn_web_development/getting_started/your_first_website/adding_interactivity/index.md

## What is JavaScript?

{{Glossary("JavaScript")}} is a fully-fledged programming language — it contains all the classic programming features you may have seen in other programming languages (or at least heard about), such as **variables**, **loops**, and **functions**.

JavaScript (when used on webpages, at least; it can also be used in other places) generally works by:

- Getting references to one or more values — such as numbers, or references to elements on the page.
- Doing something with those values, such as adding the numbers together.
- Returning a result that can be used to do something else later on. For example, you might to display the sum of those numbers on the page.

Let's look at an example — we'll return to the basic list we saw in the last couple of articles:

```html live-sample___basic-js
<p>Instructions for life:</p>

<ul>
  <li>Eat</li>
  <li>Sleep</li>
  <li>Repeat</li>
</ul>
```

> We'll also define a CSS class that styles the element it is applied to to look like it is done, with a green text color and a strikethrough:

Author: I'm not quite comfortable with this change — we need to make it clear that the CSS class will style any element it is applied to, not just list items. I've updated it to:

> We'll also define a CSS class that styles any element it is applied to (below we'll apply it to our `<li>` elements via JavaScript) to make it look like a completed task, with a green text color and a strikethrough:

WDYT?

```css live-sample___basic-js
.done {
  color: darkseagreen;
  text-decoration: line-through solid black 2px;
}
```

> Now on to the JavaScript. Here, we first store a reference to the `<li>` elements inside a variable called `listItems`. We then define a function called `toggleDone()` that adds a `class` of `done` to an element if it desn't have a class, and removes its class if it already has one. Finally, we loop through the list items (using `forEach()`) and add an event listener to each one so that when the list items are clicked, the `done` class is applied to them, which sets the CSS declared above on them.

Author: I have used most of your text suggestions from above, but I am really wary of adding too much information. I've left out the event object stuff for fear of the reader's brain melting.

3. Go to your `index.html` file and enter this code on a new line, just before the closing `</head>` tag:

   ```html
   <script async src="scripts/main.js"></script>
   ```

   This does the same job as the {{htmlelement("link")}} element for CSS. It applies the JavaScript to the page, so it can have an effect on the HTML (along with the CSS, and anything else on the page).

4. Add this code to your `scripts/main.js` file:

   ```js
   // Store a reference to the <h1> in a variable
   const myHeading = document.querySelector("h1");
   // Update the text content of the <h1>
   myHeading.textContent = "Hello world!";
   ```

5. Make sure the HTML and JavaScript files are saved, then load `index.html` in your browser. You should see something like this:

![Heading "hello world" above a firefox logo](hello-world.png)

### What happened?

We have used JavaScript to change the heading text to _Hello world!_. We grabbed a reference to the heading and stored it in a variable called `myHeading` (a container that stores a value). This is similar to what we did using CSS selectors. When you want to do something to an element, you need to select it first.

> Following that, we set the value of the `myHeading` variable's text content to _Hello world!_.

Author: We've got an issue with accuracy in this suggestion — it is not the variable that is set to "Hello world!", but the value of its `textContent` property. I've updated it to:

> Following that, we set the value of the `myHeading` variable's `textContent` property (which represents the `<h1>` element's text content) to _Hello world!_.

Does this work for you?

> Following that, we set the value of the `myHeading` variable's text content to _Hello world!_.

Reviewer: Ah, yes! Your revision looks good.

We have used JavaScript to change the heading text to _Hello world!_. We did this by using a function called {{domxref("Document.querySelector", "querySelector()")}} to grab a reference to your heading, and then store it in a variable called `myHeading`. This is similar to what we did using CSS selectors. When you want to do something to an element, you need to select it first.
We have used JavaScript to change the heading text to _Hello world!_. We grabbed a reference to the heading and stored it in a variable called `myHeading` (a container that stores a value). This is similar to what we did using CSS selectors. When you want to do something to an element, you need to select it first.

Following that, the code set the value of the `myHeading` variable's {{domxref("Node.textContent", "textContent")}} property (which represents the content of the heading) to _Hello world!_.
Following that, we set the value of the `myHeading` variable's text content to _Hello world!_.

> [!NOTE]
> Both of the features you used in this exercise are parts of the [Document Object Model (DOM) API](/en-US/docs/Web/API/Document_Object_Model), which has the capability to manipulate documents.

> ## Language basics crash course

Author: Yup, way too much info for this point.

> Before going any further, delete the "Hello world!" code from your `main.js` file. If you don't, the existing code will clash with the new code you are about to add.

Author: Yes, I agree with all of this; updates made.

1. Inside your `first-website` folder or the example folder you have just downloaded, create a new folder named `scripts`.
2. Within the `scripts` folder, create a new text document called `main.js`, and save it.
3. Go to your `index.html` file and enter this code on a new line, just before the closing `</head>` tag:

   ```html
   <script async src="scripts/main.js"></script>
   ```

   This does the same job as the {{htmlelement("link")}} element for CSS – it applies the JavaScript to the page so it can affect the HTML (along with the CSS and anything else on the page).

4. Add this code to your `scripts/main.js` file:

   ```js
   // Store a reference to the <h1> in a variable
   const myHeading = document.querySelector("h1");
   // Update the text content of the <h1>
   myHeading.textContent = "Hello world!";
   ```

5. Make sure the HTML and JavaScript files are saved, then load `index.html` in your browser. You should see something like this:

![Heading "hello world" above a firefox logo](hello-world.png)

Let's break down how this example works.

> We used JavaScript to change the heading text to `Hello world!`. We grabbed a reference to the heading and stored it in a variable called `myHeading` (a container that stores a value). This is similar to how you apply CSS to elements, selecting the elements to affect via a CSS selector before manipulating their properties. In both cases, when you want to do something to an element, you need to select it first.

Reviewer: slight tweak
```suggestion
We used JavaScript to change the heading text to `Hello world!`. We grabbed a reference to the heading and stored it in a variable called `myHeading` (a container that stores a value). This is similar to how you apply CSS to elements – you first select the elements you want to affect using a CSS selector, and then define the styles you want for those elements. In both cases, when you want to do something to an element, you need to select it first.
```

  - If it is, the code changes the `src` value to the path of the second image, forcing the other image to be loaded inside the `<img>` element.
  - If it isn't (meaning the image has already been changed), the `src` value swaps back to the original image path.

> [!NOTE]
> This section introduces several important terms. Key concepts include:
>
> - [API](/en-US/docs/Glossary/API): A set of features that allows a developer to interact with a programming environment. Web APIs (such as the DOM API features we used above) are built on top of the JavaScript language and allow you to manipulate various parts of the browser and the webpages it is displaying.
> - [Events](/en-US/docs/Learn_web_development/Core/Scripting/Events): Things that happen in the browser. They're key to making websites interactive. You can run code in response to events using **event handler functions** – these are code blocks that run when an event occurs. The most common example is the [click event](/en-US/docs/Web/API/Element/click_event), which is fired by the browser when a user clicks on something.

> > - [Functions](/en-US/docs/Learn_web_development/Core/Scripting/Functions): A way of packaging code that you wish to reuse. You can define your code once inside a function and then run it as many times as you like, which helps you avoid writing the same code over and over. In the above code we defined a `click` event handler function, which is run every time a user clicks the image.

Reviewer: ```suggestion
> - [Functions](/en-US/docs/Learn_web_development/Core/Scripting/Functions): A way of packaging code that you wish to reuse. You can define your code inside a function once and then run it as many times as you like, which helps you avoid writing the same code over and over. In our example here, we defined a `click` event handler function, which runs every time a user clicks the image.
```

4. Save all files and load `index.html` in the browser. Now when you click the image, it should change to the other one.

In the above code, you stored a reference to your {{htmlelement("img")}} element in `myImage`. Next, you gave it a `click` event handler function. Every time the `<img>` is clicked, the function:

1. Retrieves the value of the image's `src` attribute.
2. Uses a conditional (`if ... else` structure) to check if the `src` value is equal to the path of the original image:

   1. If it is, the code changes the `src` value to the path of the second image, forcing the other image to be loaded inside the {{htmlelement("img")}} element.
   2. If it isn't (meaning it must already have changed), the `src` value swaps back to the original image path.

> [!NOTE]
> This section contains several items of important terminology. The key terms are:
>

> > - [API](/en-US/docs/Glossary/API): A set of functionality that allows a developer to interact with a programming environment. WebAPIs (such as the DOM API features we used above) are built on top of the JavaScript language and allow you to manipulate various parts of the browser.

Author: Your suggestions are good, so I've used them, but I wasn't happy with my sentence end — it isn't just parts of the browser that Web APIs manipulate. I've edited it to this:

> - [API](/en-US/docs/Glossary/API): A set of features that allows a developer to interact with a programming environment. Web APIs (such as the DOM API features we used above) are built on top of the JavaScript language and allow you to manipulate various parts of the browser and the webpages it is displaying.

> - [Events](/en-US/docs/Learn_web_development/Core/Scripting/Events): Real interactivity on a website requires events — things that happen in the browser. You can run code in response to events using **event handler functions**, code structures that run code in response to an event happening. The most obvious example is the [click event](/en-US/docs/Web/API/Element/click_event), which is fired by the browser when you click on something with your mouse.

> > - [Functions](/en-US/docs/Learn_web_development/Core/Scripting/Functions): A way of packaging code that you wish to reuse. You can define your code inside a function and execute it when you call the function name in your code, avoiding having to repeatedly write the same code over and over again.

Author: "calling the function's name" sounds a bit odd and might confuse people because in the above example, we use an anonymous function rather than a named function. I've updated this bullet to:

> - [Functions](/en-US/docs/Learn_web_development/Core/Scripting/Functions): A way of packaging code that you wish to reuse. You can define your code once inside a function and then run it as many times as you like, which helps you avoid writing the same code over and over. In the above code we defined a `click` event handler function, which is run every time a user clicks the image.

Does that work for you?

> > - [Functions](/en-US/docs/Learn_web_development/Core/Scripting/Functions): A way of packaging code that you wish to reuse. You can define your code inside a function and execute it when you call the function name in your code, avoiding having to repeatedly write the same code over and over again.

Reviewer: Agree! This rephrasing is better.

   }
   ```

>    The `setUserName()` function contains a [`prompt()`](/en-US/docs/Web/API/Window/prompt) function, which displays a dialog box, similar to `alert()`. This `prompt()` function does more than `alert()`, asking the user to enter data, and storing it in a variable after the user clicks _OK_. In this case, we are asking the user to enter a name. Next, the code calls on the `localStorage` API, which allows us to store data in the browser and retrieve it later. We use the {{domxref("Storage.setItem", "localStorage.setItem()")}} function to create and store a data item called `"name"`, setting its value to the `myName` variable, which contains the user's entry for the name. Finally, we set the `textContent` of the heading to a string with the user's newly-stored name added to the end of it.

Reviewer: - okay to link alert() to https://developer.mozilla.org/en-US/docs/Web/API/Window/alert?
- we could add para breaks at "Next" and "Finally"
- can we also add a link for "`localStorage` API"?
```suggestion
   The `setUserName()` function contains a [`prompt()`](/en-US/docs/Web/API/Window/prompt) function, which displays a dialog box similar to `alert()`. This `prompt()` function does more than alerting; it asks the user to enter data and stores it in a variable after the user clicks _OK_. In this example, we're asking the user to enter a name and storing it in `myName`.
  
  Next, the code uses the `localStorage` API, which allows us to store data in the browser and retrieve it later. We use the {{domxref("Storage.setItem", "localStorage.setItem()")}} function to create and store a data item called `"name"`, setting its value to the `myName` variable, which contains the user's input.
  
  Finally, we set the `textContent` of the heading to a string that includes the user's stored name.
```

Author: I like your suggestion here, so I've used all of it, but with a couple of tweaks:

1. I removed all mention of `alert()` because we are not using it anywhere in the article, and the way it is written makes it sound like we've used it previously.
2. I've added a link to "Web Storage API" rather than "`localStorage` API", as this is the real landing page that sits over the top of all this functionality.

I ended up with:

> The `setUserName()` function contains a [`prompt()`](/en-US/docs/Web/API/Window/prompt) function, which asks the user to enter data and stores it in a variable after they click _OK_. In this example, we're asking the user to enter a name and storing it in `myName`.
>
> Next, the code uses the [Web Storage API](/en-US/docs/Web/API/Web_Storage_API), which allows us to store data in the browser and retrieve it later. We use the {{domxref("Storage.setItem", "localStorage.setItem()")}} function to create and store a data item called `"name"`, setting its value to the `myName` variable, which contains the user's input.
>
> Finally, we set the `textContent` of the heading to a string that includes the user's stored name.

Author: One issue I've got is that I'm finding it impossible to get the above multiple paragraphs to render with line breaks separating them when nested in the ordered list, without using double `<br />` tags at the end of each paragraph.

It works, but it is really crappy markup

Reviewer: This version looks perfect. Thanks!

> 4. Add the following condition block after the function declaration. This is our _initialization code_: it runs the functionality when the page first loads:

Reviewer: ```suggestion
4. Add the following conditional block after the function declaration. This is our _initialization code_ – it runs when the page first loads:
```

Author: I've used your suggestion, but added a bit to the end:

> 4. Add the following conditional block after the function declaration. This is our _initialization code_ — it runs when the page first loads to start the program off:

The point of initialization code is that it starts the program running when the page first loads. It is the first domino in the cascade, as it were.

Reviewer: That makes sense

   }
   ```

>    The first line of this block uses the negation operator (logical NOT, represented by the `!` character) to check whether the `name` data item is _not_ already stored in `localStorage`. If not, the `setUserName()` function runs to create it. If it exists (that is, the user set a username during a previous visit), we retrieve the stored name using {{domxref("Storage.getItem", "localStorage.getItem()")}} and set the `textContent` of the heading to a string, plus the user's name, as we did inside `setUserName()`.

Reviewer: ```suggestion
   The first line of this block uses the negation operator (logical NOT, represented by the `!` character) to check whether the `name` data item is _not_ already stored in `localStorage`. If not, the `setUserName()` function runs to create it. If it exists (that is, the user set a username during a previous visit), we retrieve the stored name using {{domxref("Storage.getItem", "localStorage.getItem()")}} and set the `textContent` of the heading to a string, plus the user's name – just like we did inside `setUserName()`.
```

> 5. Add a `click` event handler function to the button. When clicked, `setUserName()` runs. This allows the user to store a different name if wished.

Reviewer: ```suggestion
5. Add a `click` event handler function to the button. When clicked, `setUserName()` runs. This allows the user to store a different name if they want to.
```

   ```js
   myButton.addEventListener("click", () => {
     setUserName();
   });
   ```

> 6. Save all files and load `index.html` in the browser. You should immediately be asked to enter your name. When you do so, it will appear inside the `<h1>` as part of the personalized greeting. Also, note how the personalization persists when you reload the page, and try pressing the "Change user" button and entering a new name.

Reviewer: ```suggestion
6. Save all files and load `index.html` in the browser. You should immediately be asked to enter your name. After you do so, it will appear inside the `<h1>` as part of the personalized greeting. Notice how the personalization persists even after you reload the page. You can click the "Change user" button to enter a new name.
```

> [!NOTE]

> > The term [operator](/en-US/docs/Learn_web_development/Core/Scripting/Math) is a general term referring to a JavaScript language character that carries out an operation on a value or values. Examples include `+` (adds two values together), `-` (subtracts one value from another), and `!` (negates a value — you met this above).

Reviewer: ```suggestion
> The term [operator](/en-US/docs/Learn_web_development/Core/Scripting/Math) refers to a JavaScript language character that carries out an operation on one or more values. Examples include `+` (adds values), `-` (subtracts one value from another), and `!` (negates a value — as you saw earlier).
```

> ### A user name of null?

Reviewer: lone h3. should we change this to an h2
```suggestion
### Handling empty input
```

Author: I agree with promoting it to an H2, but I wanted to keep the title. This section isn't just about empty inputs, it is about ending up with a null value too, and I'd rather mention the null case in the title, as the empty case is a bit more intuitive.

I could mention them both, something like "Handling empty and null inputs", but that feels a bit long, and also, you could argue that the null case isn't really an input (it is what happens when the user cancels the input...)

Reviewer: Subtle nuance but all important points. I see what you mean and I agree.

Suggesting another alternative to keep and address the "null" case

> When you run the example and get the dialog box that prompts you to enter your user name, try pressing the _Cancel_ button. You should end up with a title that reads _Mozilla is cool, null_. This happens because the value is set as [`null`](/en-US/docs/Web/JavaScript/Reference/Operators/null) when you cancel the prompt. _null_ is a special value in JavaScript that refers to the absence of a value.

Reviewer: ```suggestion
When you run the example and get the dialog box that prompts you to enter your name, try pressing the _Cancel_ button. You should end up with a title that reads _Mozilla is cool, null_. This happens because the value is set to [`null`](/en-US/docs/Web/JavaScript/Reference/Operators/null) when you cancel the prompt. In JavaScript, _null_ is a special value that represents the absence of a value.
```

Also, try clicking _OK_ without entering a name. You should end up with a title that reads _Mozilla is cool,_ because you've set `myName` to an empty string.

> To avoid these problems, you could check that the user hasn't entered a blank name using another conditional. Update your `setUserName()` function to this:

Reviewer: ```suggestion
To avoid these problems, you can add another conditional to check that the user hasn't entered a blank name. Update your `setUserName()` function to the following:
```

5. Add a `click` event handler function to the button. When clicked, `setUserName()` runs. This allows the user to store a different name if they want to.

   ```js
   myButton.addEventListener("click", () => {
     setUserName();
   });
   ```

6. Save all files and load `index.html` in the browser. You should immediately be asked to enter your name. After you do so, it will appear inside the `<h1>` as part of the personalized greeting. Notice how the personalization persists even after you reload the page. You can click the "Change user" button to enter a new name.

> [!NOTE]
> The term [operator](/en-US/docs/Learn_web_development/Core/Scripting/Math) refers to a JavaScript language character that carries out an operation on one or more values. Examples include `+` (adds values), `-` (subtracts one value from another), and `!` (negates a value — as you saw earlier).

> ## A user name of null?

Reviewer: WDYT about:
```suggestion
## Avoiding a null user name
```

Author: This still doesn't handle the empty case. I'm just going to keep it as-is.

If you get stuck, you can compare your work with our [finished example code on GitHub](https://github.com/mdn/beginner-html-site-scripted/blob/main/scripts/main.js).

> Here, we have only really scratched the surface of JavaScript. You'll learn a lot more in our [Dynamic scripting with JavaScript](/en-US/docs/Learn_web_development/Core/Scripting) Core module later on in the course.

Reviewer: ```suggestion
We've only really scratched the surface of JavaScript in this article. You'll learn a lot more in our [Dynamic scripting with JavaScript](/en-US/docs/Learn_web_development/Core/Scripting) Core module later on in the course.
```
